overlord/ifacestate/hotplug.go

```diff
@@ -84,11 +106,275 @@ func defaultDeviceKey(devinfo *hotplug.HotplugDeviceInfo, keyVersion int) (strin
+			return
+		}
+
+		// if we know this slot already, check if its static attributes changed - if so, we need to update the repo and connections (if any)
+		slot, err := m.repo.SlotForHotplugKey(iface.Name(), key)
+		if err != nil {
+			logger.Noticef("internal error: %s", err)
+		}
+
+		// Add or update slot in the repository
+		if slot != nil {
+			if reflect.DeepEqual(slotSpec.Attrs, slot.Attrs) {
+				// slot attributes unchanged, nothing to do
+				logger.Debugf("Slot %s for device %s already present and unchanged", slot.Name, key)
+			} else {
+				logger.Debugf("Slot %s for device %s has changed", slot.Name, key)
+				ts := updateDevice(st, iface.Name(), key, slotSpec.Attrs)
+				chg := st.NewChange(fmt.Sprintf("hotplug-update-%s", iface), fmt.Sprintf("Update hotplug slot of interface %s, device %s", iface.Name(), key))
+				chg.AddAll(ts)
+			}
+		} else {
+			// Determine slot name:
+			// - if a slot for given hotplug key exists in hotplug state, use old name
+			// - otherwise use name provided by slot spec, if set
+			// - if not, use auto-generated name.
+			proposedName := slotSpec.Name
+			for _, stateSlot := range stateSlots {
+				if stateSlot.HotplugKey == key {
+					proposedName = stateSlot.Name
+					break
+				}
+			}
+			if proposedName == "" {
+				proposedName = suggestedSlotName(devinfo, iface.Name())
+			}
+			proposedName = ensureUniqueName(proposedName, func(name string) bool {
+				if slot, ok := stateSlots[name]; ok {
+					return slot.HotplugKey == key
+				}
+				return m.repo.Slot(coreSnapInfo.InstanceName(), name) == nil
+			})
+			newSlot := &snap.SlotInfo{
+				Name:       proposedName,
+				Label:      slotSpec.Label,
+				Snap:       coreSnapInfo,
+				Interface:  iface.Name(),
+				Attrs:      slotSpec.Attrs,
+				HotplugKey: key,
+			}
+			if iface, ok := iface.(interfaces.SlotSanitizer); ok {
+				if err := iface.BeforePrepareSlot(newSlot); err != nil {
+					logger.Noticef("Failed to sanitize hotplug-created slot %q for interface %s: %s", newSlot.Name, newSlot.Interface, err)
+					continue
+				}
+			}
+
+			if err := m.repo.AddSlot(newSlot); err != nil {
+				logger.Noticef("Failed to create slot %q for interface %s: %s", newSlot.Name, newSlot.Interface, err)
+				continue
+			}
+			stateSlots[newSlot.Name] = &HotplugSlotInfo{
+				Name:        newSlot.Name,
+				Interface:   newSlot.Interface,
+				StaticAttrs: newSlot.Attrs,
+				HotplugKey:  newSlot.HotplugKey,
+			}
+			setHotplugSlots(st, stateSlots)
+
+			logger.Noticef("Added hotplug slot %s:%s of interface %s for hotplug key %q", newSlot.Snap.InstanceName(), newSlot.Name, newSlot.Interface, key)
+
+			chg := st.NewChange(fmt.Sprintf("hotplug-connect-%s", iface), fmt.Sprintf("Connect hotplug slot of interface %s", iface.Name()))
+			hotplugConnect := st.NewTask("hotplug-connect", fmt.Sprintf("Recreate connections of device %q", key))
+			setHotplugAttrs(hotplugConnect, iface.Name(), key)
+			chg.AddTask(hotplugConnect)
+		}
+		st.EnsureBefore(0)
```

Comment:
Should this be triggered for each interface and even if feature is disabled?

Reply:
See the earlier explanation. It's only triggered for hotplug-aware interfaces, we only let the interfaces create slot specifications and device keys, but these operations have no side effects (we check the feature flag early in the loop).

I'm of course happy to revisit the original idea of doing some of these things even if feature is disabled, in which cause we would just have udev monitor active and reporting events, and just discard them immediately.